core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -640,13 +642,27 @@ private synchronized void checkTaskCompletion()
         outputBuffer.setNoMorePages();
 
         BufferState bufferState = outputBuffer.getState();
-        if (bufferState != BufferState.FINISHED) {
+        if (!bufferState.isTerminal()) {
             taskStateMachine.transitionToFlushing();
             return;
         }
 
-        // Cool! All done!
-        taskStateMachine.finished();
+        if (bufferState == BufferState.FINISHED) {
+            // Cool! All done!
+            taskStateMachine.finished();
+            return;
+        }
+
+        if (bufferState == BufferState.FAILED) {
+            Throwable failureCause = outputBuffer.getFailureCause()
+                    .orElseGet(() -> new TrinoException(GENERIC_INTERNAL_ERROR, "Output buffer is failed but the failure cause is missing"));
+            taskStateMachine.failed(failureCause);
+            return;
+        }
+
+        // The only terminal state that remains is ABORTED.
+        // Buffer is expected to be aborted only if the task itself is aborted. In this scenario the following statement is expected to be noop.
```

Comment:
`In this scenario the following statement is expected to be noop.`

why? because task is aborted so this line should never execute?

Reply:
Failing an aborted task is a noop, as the `ABORTED` state is a terminal state.